### PR TITLE
MOLDEN: output frequency information

### LINKED ISSUE #1127: MOLDEN: output frequency information
Hi there, 

I need a molden file to generate a wigner sampling esamble. I was wondering if cclib could generate it. It has to contain geometry, frequency modes and displacement vector in it. 

Orca v. 5.0.3

### PR DESCRIPTION
**Pull request by berquist**
Closes #1127

### CODE REVIEW

File: cclib/io/moldenwriter.py [hunk 1]
```diff
@@ -32,7 +33,11 @@ def round_molden(num, p=6):
 
 
 class MOLDEN(filewriter.Writer):
-    """A writer for MOLDEN files."""
+    """A writer for MOLDEN files.
+
+    Documentation of the format is located at
+    https://www3.cmbi.umcn.nl/molden/molden_format.html.
```
Comment (shivupa) [inline cclib/io/moldenwriter.py:39]: nice

File: cclib/io/moldenwriter.py [hunk 2]
```diff
@@ -197,6 +202,54 @@ def _mo_from_ccdata(self):
...
+            vibfreqs_lines.extend([f"{vibfreq:16.8f}" for vibfreq in vibfreqs])
+            lines.append("\n".join(vibfreqs_lines))
+
+        if hasattr(self.ccdata, "atomcoords") and hasattr(self.ccdata, "atomnos"):
+            atomcoords = utils.convertor(self.ccdata.atomcoords[0], "Angstrom", "bohr")
+            atomsyms = (self.pt.element[atomno] for atomno in self.ccdata.atomnos)
+            atomcoords_lines = ["[FR-COORD]"]
+            for atomsym, atomcoord in zip(atomsyms, atomcoords):
+                atomcoords_lines.append(
+                    "{:3s} {:15.8f} {:15.8f} {:15.8f}".format(atomsym, *atomcoord)
```
Comment (shivupa) [inline cclib/io/moldenwriter.py:221]: can we switch to fstrings (used below)

File: cclib/io/moldenwriter.py [hunk 3]
```diff
@@ -197,6 +202,54 @@ def _mo_from_ccdata(self):
...
+            lines.append("\n".join(atomcoords_lines))
+
+        if hasattr(self.ccdata, "vibdisps"):
+            vibdisps = self.ccdata.vibdisps
+            vibdisps_lines = ["[FR-NORM-COORD]"]
+            for vibidx in range(vibdisps.shape[0]):
+                vibdisps_lines.append(f"vibration {vibidx + 1}")
+                for iatom in range(vibdisps.shape[1]):
+                    vibdisps_lines.append(
+                        "{:15.8f} {:15.8f} {:15.8f}".format(*vibdisps[vibidx, iatom])
```
Comment (shivupa) [inline cclib/io/moldenwriter.py:232]: can we switch to fstrings (used below)

File: cclib/io/moldenwriter.py [hunk 4]
```diff
@@ -197,6 +202,54 @@ def _mo_from_ccdata(self):
...
+            atomsyms = (self.pt.element[atomno] for atomno in self.ccdata.atomnos)
+            atomcoords_lines = ["[FR-COORD]"]
+            for atomsym, atomcoord in zip(atomsyms, atomcoords):
+                atomcoords_lines.append(
+                    "{:3s} {:15.8f} {:15.8f} {:15.8f}".format(atomsym, *atomcoord)
+                )
+            lines.append("\n".join(atomcoords_lines))
+
+        if hasattr(self.ccdata, "vibdisps"):
+            vibdisps = self.ccdata.vibdisps
```
Comment (shivupa) [inline cclib/io/moldenwriter.py:226]: do these not need converted to delta Bohr?

**Review by shivupa — CHANGES_REQUESTED** [review]

File: cclib/io/moldenwriter.py (hunk 4 above)
Comment (berquist) [reply cclib/io/moldenwriter.py:221]: They do, I missed it.

**Comment by berquist** [comment]
Marked as draft because producing vibdisps in bohr doesn't match Q-Chem MOLDEN output, which I think is a Q-Chem bug, but I can't find another program that prints `[FR-NORM-COORD]` in its output (have only tried ORCA and DALTON). I want to add a cclib-generated MOLDEN file with frequency information that matches a known good source first.

**Comment by ghutchis** [comment]
From https://www3.cmbi.umcn.nl/molden/molden_format.html
>Specification of frequencies and corresponding normal coordinates.
>The atomic coordinates x,y,z and atomic displacements dx,dy,dz are all in Bohr (Atomic Unit of length).

**Comment by berquist** [comment]
After managing to get identical MOLDEN output from Psi4 and looking at LiH frequency outputs from a few programs, I think the problem is actually understanding what units `vibdisps` is in.  `data_notes.rst` doesn't say, but the `ccData` docstring says delta angstrom, which I think is wrong.  The parsers save `vibdisps` as-is, and when you look at the outputs pretty much none say what the units are, except for DALTON `((bohrs*amu**(1/2))` and ORCA (kind of: `These modes are the Cartesian displacements weighted by the diagonal matrix M(i,i)=1/sqrt(m[i]) where m[i] is the mass of the displaced atom`).  

| program  | freq (wavenumber) | disp Li   | disp H   |
|----------|-------------------|-----------|----------|
| DALTON   | 1334.71           | -0.1338   | 0.931454 |
| Gaussian | 1334.7151         | -0.14219  | 0.98984  |
| NWChem   | 1334.715          | -0.13380  | 0.93145  |
| ORCA     | 1335.43           | -0.143716 | 0.989619 |
| Q-Chem   | 1334.72           | -0.142    | 0.99     |

Even without going to diagonalize the Hessians myself to see if the two groupings come from the mass-weighting, it very much seems like we are presenting `vibdisps` in atomic units and the docstring is wrong.

**Comment by shivupa** [comment]
Thanks for diagnosing that. Can you update the docstring https://github.com/cclib/cclib/blob/592a8710704e97a1e17179d6f2d27cc5076bab9a/cclib/parser/data.py#L90 and https://github.com/cclib/cclib/blob/master/doc/sphinx/data_notes.rst

This is ready to go following that

**Review by shivupa — APPROVED** [review]

**Comment by berquist** [comment]
> Thanks for diagnosing that. Can you update the docstring
> 
> https://github.com/cclib/cclib/blob/592a8710704e97a1e17179d6f2d27cc5076bab9a/cclib/parser/data.py#L90
> and [https://github.com/cclib/cclib/blob/master/doc/sphinx/data_notes.rst](https://github.com/cclib/cclib/blob/master/doc/sphinx/data_notes.rst?rgh-link-date=2022-07-08T16%3A49%3A17Z)
> 
> This is ready to go following that

I need to try and find out if all are mass-weighted first. I wonder if that accounts for the two "groups" of displacements.

**Review by amandadumi — COMMENTED** [review]
So far this all looks good, will keep this as a general comment until the mass-weighted stuff is sorted, thank you for looking into that, it's tricky without documentation for sure.

**Comment by berquist** [comment]
I found code in Psi4 that will print normal modes in all reasonable forms and hacked it to get:

| description                                                                | disp Li | disp H |
|----------------------------------------------------------------------------|---------|--------|
| $q$ `[a0 u^1/2]`, the mass-weighted normalized eigenvectors of the Hessian | -0.3544 | 0.9351 |
| $w$ `[a0]`, the un-mass-weighted (Cartesian) of q                          | -0.1338 | 0.9315 |
| $x$ `[a0]`, the normalized w                                               | -0.1422 | 0.9898 |

The breakdown for default printing is then:

| program  | normal mode printing? |
|----------|-----------------------|
| DALTON   | $w$                   |
| GAMESS   | $w$                   |
| Gaussian | $x$                   |
| NWChem   | $w$                   |
| ORCA     | $x$                   |
| Psi4     | $x$                   |
| Q-Chem   | $x$                   |

so all programs are taking eigenvectors of the mass-weighted Hessian and then un-mass-weighting them, then some are renormalizing them?

What we put as the docstring depends on whether or not we want to
- only present $x$ and convert from $w$,
- only present $w$ and convert from $x$, or
- present exactly what is parsed with the documented knowledge that not all programs return (re-)normalized modes.

Regarding MOLDEN output specifically, it says nothing about this, but we can probably assume that what goes there is exactly what gets printed in the main output file.

---

Unrelated, but aren't the units of $x$ according to Psi4 wrong (https://github.com/psi4/psi4/blob/8c6e5bfa5c501567d7ec1ee3d4c20b26f814e90c/psi4/driver/qcdb/vib.py#L586 and in many other comments/docstrings)?  The DALTON comment about units is what I'd expect.  If others agree, I will ask Lori.

**Comment by langner** [comment]
This roughly still looks good to go in. @shivupa is your approval still fresh? @amandadumi do you have any concerns/comments?

**Review by shivupa — APPROVED** [review]
I am still happy with this

**Comment by shivupa** [comment]
> Unrelated, but aren't the units of
according to Psi4 wrong (https://github.com/psi4/psi4/blob/8c6e5bfa5c501567d7ec1ee3d4c20b26f814e90c/psi4/driver/qcdb/vib.py#L586 and in many other comments/docstrings)? 

It does seem that way.

**Pull request merged**